pkg/ibmcsidriver/controller.go

```diff
@@ -127,6 +127,112 @@ func (csiCS *CSIControllerServer) CreateVolume(ctx context.Context, req *csi.Cre
+		any securityGroupID then
+
+		1. IKS cluster security group is fetched and used by CSI driver
+		2. Else default VPC security group is considerd by the VPC IAAS layer.
+
+	*/
+
+	if requestedVolume.AccessControlMode == SecurityGroup {
+		/* Skip GetSubnetForVolumeAccessPoint call if user has not provided SubnetID but PrimaryIPID is provided.
+		For all rest of the following use cases if subnetId is not provided we fetch subnet
+		1.) User has not provided anything just ENI/VNI is enabled (Any random IP Address will be created and assigned to VNI in the fetch subnet range)
+		2.) User has provided PrimaryIP Address. (The respective IP Address will be created and assigned to VNI in the fetched subnet range)
+		*/
+		subnetID := requestedVolume.SubnetID
+
+		if len(subnetID) == 0 && (requestedVolume.PrimaryIP == nil || len(requestedVolume.PrimaryIP.ID) == 0) {
+			subnetIDList := os.Getenv("VPC_SUBNET_IDS")
+
+			//We need to abort here as there is no use of going ahead and fetching the matching subnet with empty list
+			if len(subnetIDList) == 0 {
+				return nil, commonError.GetCSIError(ctxLogger, commonError.SubnetIDListNotFound, requestID, nil)
+			}
+
+			subnetReq := provider.SubnetRequest{
+				SubnetIDList:  subnetIDList,
+				ZoneName:      requestedVolume.Az,
+				VPCID:         os.Getenv("VPC_ID"),
+				ResourceGroup: requestedVolume.ResourceGroup,
+			}
+
+			ctxLogger.Info("Getting Subnet for VolumeAccessPoint...")
+
+			subnetID, err = session.GetSubnetForVolumeAccessPoint(subnetReq)
+			if err != nil || len(subnetID) == 0 {
+				return nil, commonError.GetCSIError(ctxLogger, commonError.SubnetFindFailed, requestID, err, requestedVolume.Az, subnetIDList)
+			}
+
+			requestedVolume.SubnetID = subnetID
+			ctxLogger.Info("Subnet fetched for VolumeAccessPoint", zap.Reflect("subnetID", subnetID))
+		}
+
+		//If securityGroup parameter is not populated via storage class
+		if requestedVolume.SecurityGroups == nil {
+			securityGroupReq := provider.SecurityGroupRequest{
+				Name:          "kube-" + csiCS.CSIProvider.GetClusterID(),
+				VPCID:         os.Getenv("VPC_ID"),
+				ResourceGroup: requestedVolume.ResourceGroup,
+			}
+
+			ctxLogger.Info("Getting SecurityGroup for VolumeAccessPoint...")
+
+			securityGroupID, err := session.GetSecurityGroupForVolumeAccessPoint(securityGroupReq)
+			if err != nil || len(securityGroupID) == 0 {
+				// If IKS Cluster SG is not available pass empty SG. VPC IAAS will consider VPC Default SG.
+				ctxLogger.Warn("SecurityGroup find failed for VolumeAccessPoint.VPC default SG will be considered", zap.Error(err))
+			} else {
+				requestedVolume.SecurityGroups = &[]provider.SecurityGroup{
+					{
+						ID: securityGroupID,
+					},
+				}
+				ctxLogger.Info("SecurityGroup fetched for VolumeAccessPoint", zap.Reflect("securityGroupID", securityGroupID))
+			}
+		}
+	} else { // IF VPC Mode
+		requestedVolume.VPCID = os.Getenv("VPC_ID")
+	}
+
```

Comment:
As of now there is no functional impact of this condition. Even if do this check it will go ahead and just check volumeAccessPoint status is pending or stable. In all cases if CreateVolume is success it is 100% guaranteed that VolumeAccessPoint is created otherwise CreateVolume as whole will fail. So there should be always volumeAccessPoint for our PVC volume. This will go ahead and check for status of access point and take action and k8s will retry on top of library retries if accesspoint status is not valid